packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts

```diff
@@ -192,6 +192,46 @@ export class SummarizerNode implements IRootSummarizerNode {
 		return { summary: result.summary, stats: result.stats };
 	}
 
+	public validateSummary() {
+		this.validateSummaryCore(false);
+	}
+
+	protected validateSummaryCore(parentSkipRecursion: boolean) {
+		assert(
+			this.wipSummaryLogger !== undefined,
+			0x1a3 /* "wipSummaryLogger should have been set in startSummary or ctor" */,
+		);
+		assert(this.wipReferenceSequenceNumber !== undefined, 0x1a4 /* "Not tracking a summary" */);
+		let didNodeSummarize = this.wipLocalPaths !== undefined;
+
+		if (parentSkipRecursion) {
+			if (this._latestSummary !== undefined) {
+				didNodeSummarize = true;
+			} else {
+				return;
+			}
+		}
+
+		/**
+		 * The absence of wip local path indicates that summarize was not called for this node. This can happen if:
+		 * 1. A child node was created after summarize was already called on the parent. For example, a data store
+		 * is realized (loaded) after summarize was called on it creating summarizer nodes for its DDSes. In this case,
+		 * parentSkipRecursion will be true and the if block above would handle it.
+		 * 2. A new node was created but summarize was never called on it. This can mean that the summary that is
+		 * generated may not have the data from this node. We should not continue, log and throw an error. This
+		 * will help us identify these cases and take appropriate action.
+		 */
+		if (!didNodeSummarize) {
+			this.throwUnexpectedError({
```

Comment:
>throwUnexpectedError [](http://example.com/codeflow?start=8&length=20)

I wonder if we'd rather simply return false and never throw during validation.

Reply:
I'd vote for returning a strongly-typed object - maybe even the exact error object currently being thrown (but with typings for the interesting bits like `retryAfterSeconds`). Especially looking at the callsite in `containerRuntime`, which wraps this single function call in try-catch just to get the error out.

Eh, at the same time, it's a smaller change and matches the way other stages are tracked with the current code.  So no disrespect if you keep it as-is.

Reply:
Added this, please check.